### PR TITLE
Switch from Coveralls to Codecov

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 okay, I suck at this CI stuff. Can you see anything obvious on what I'm doing wrong? The Codecov action doesn't seem to get run at all.

**Comment by superm1** [comment]
> Error: Codecov token not found. Please provide Codecov token with -t flag.

Did you add that token?

**Comment by superm1** [comment]
Well you got me really confused, the syntax looks right according to https://github.com/codecov/codecov-action.  Did you copy/paste the token properly when you saved the secret (its not possible to view a secret after it's saved, so I don't know).

**Comment by hughsie** [comment]
I put it in the `Repository secrets` section with a name of `CODECOV_TOKEN` and then pasted in the GUID from codecov. I've just pasted it in again and saved just to be on the safe side.

**Comment by hughsie** [comment]
Ohh I wonder if we have to regenerate the containers with the new secret included?

**Comment by superm1** [comment]
No, the secrets don't go into the container. They're in the environment

**Comment by superm1** [comment]
No, the secrets don't go into the container. They're in the environment

**Comment by superm1** [comment]
It might be this: https://github.com/codecov/codecov-action/issues/1463

Add the secret to dependabot and see if that works.

**Comment by hughsie** [comment]
I'd concur if this was a dependabot-generated PR -- but this isn't. I'm wondering if I should have added the secret to the fwupd organisation rather than the fwupd/fwupd project -- although I don't see why that would make any difference.

**Comment by superm1** [comment]
> I'd concur if this was a dependabot-generated PR -- but this isn't. I'm wondering if I should have added the secret to the fwupd organisation rather than the fwupd/fwupd project -- although I don't see why that would make any difference.

Oh hmm.

**Comment by hughsie** [comment]
Latest theory: https://stackoverflow.com/questions/78298827/why-is-codecov-upload-step-in-github-actions-not-finding-the-token

File: .github/workflows/matrix.yml [hunk 1]
```diff
@@ -43,22 +47,13 @@ jobs:
...
-        uses: coverallsapp/github-action@4cdef0b2dbe0c9aa26bed48edb868db738625e79 # v2.3.3
+      - name: Codecov
+        if: matrix.os == 'debian-x86_64'
+        uses: codecov/codecov-action@v4
         with:
-          parallel-finished: true
-          carryforward: "run-debian-i386,run-debian-x86_64"
+          fail_ci_if_error: true # optional (default = false)
+          verbose: true # optional (default = false)
+          token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (superm1) [inline .github/workflows/matrix.yml:56]: I think try to make this `secrets.codecov_token`

**Comment by superm1** [comment]
I've got it working (uploading a report).  It won't actually submit the report until it's merged to main.  So for now I want to keep coveralls enabled and we can back it out once we're sure we're happy with codecov.

**Comment by hughsie** [comment]
@superm1 many thanks for doing this! Did you have any idea why the Coveralls and Codecov values are so different?

**Comment by superm1** [comment]
> @superm1 many thanks for doing this! Did you have any idea why the Coveralls and Codecov values are so different?

Well they both take the exact same XML, so they must be interpreting the results differently.  Maybe they give a different weighting to branches that only have partial coverage?

**Pull request merged**